langchain4j-clickhouse/src/main/java/dev/langchain4j/store/embedding/clickhouse/ClickHouseEmbeddingStore.java

```diff
@@ -0,0 +1,310 @@
+        }
+
+        public Builder url(String url) {
+            settings.setUrl(url);
+            return this;
+        }
+
+        public Builder username(String username) {
+            settings.setUsername(username);
+            return this;
+        }
+
+        public Builder password(String password) {
+            settings.setPassword(password);
+            return this;
+        }
+
+        public Builder database(String database) {
+            settings.setDatabase(database);
+            return this;
+        }
+
+        public Builder table(String table) {
+            settings.setTable(table);
+            return this;
+        }
+
+        public Builder dimension(int dimension) {
+            settings.setDimension(dimension);
+            return this;
+        }
+
+        /**
+         * Column type map to project column name onto langchain semantics.
+         * <p>Must have keys: `text`, `id`, `embedding`</p>
+         * <p>Optional key: metadata</p>
+         *
+         * @param columnMap column map
+         * @return builder
+         */
+        public Builder columnMap(Map<String, String> columnMap) {
+            settings.setColumnMap(columnMap);
+            return this;
+        }
+
+        public Builder properties(Properties properties) {
+            settings.setProperties(properties);
+            return this;
+        }
+
+        public Builder setting(ClickHouseSettings setting) {
+            this.settings = setting;
+            return this;
+        }
+
+        public ClickHouseEmbeddingStore build() {
+            return new ClickHouseEmbeddingStore(dataSource, settings);
+        }
+    }
+
+    @Override
+    public String add(Embedding embedding) {
+        String id = randomUUID();
+        add(id, embedding);
+        return id;
+    }
+
+    @Override
+    public void add(String id, Embedding embedding) {
+        addInternal(id, embedding, null);
+    }
+
+    @Override
+    public String add(Embedding embedding, TextSegment textSegment) {
+        String id = randomUUID();
+        addInternal(id, embedding, textSegment);
+        return id;
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings) {
+        List<String> ids = embeddings.stream()
+                .map(ignored -> randomUUID())
+                .collect(toList());
+        addAllInternal(ids, embeddings, null);
+        return ids;
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedded) {
+        List<String> ids = embeddings.stream()
+                .map(ignored -> randomUUID())
+                .collect(toList());
+        addAllInternal(ids, embeddings, embedded);
+        return ids;
+    }
+
+    @Override
+    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
+        try (Connection conn = createConnection();
+             Statement stmt = conn.createStatement();
```

Comment:
The statement has to be closed. Probably the best way is to wrap it with try with resources block

Reply:
It seems that it's already in try-with-resources block.

Reply:
It's not. Connection ("conn") will be closed, statement won't close automatically.  So you'll have a leak.

Reply:
But why? Statement is already in try-with-resources block.

```java
try (Connection conn = createConnection();
       Statement stmt = conn.createStatement()) {
    // ...
}

Reply:
My bad, stupid syntax, didn't see the brackets correctly. Sorry.